crates/miden-objects/src/address/routing_parameters.rs

```diff
@@ -417,6 +257,150 @@ impl Deserializable for RoutingParameters {
+fn encode_receiver_profile(interface: AddressInterface, note_tag_len: Option<u8>) -> [u8; 2] {
+    let note_tag_len = note_tag_len.unwrap_or(ABSENT_NOTE_TAG_LEN);
+
+    let interface = interface as u16;
+    debug_assert_eq!(interface >> 11, 0, "address interface should have its upper 5 bits unset");
+
+    // The interface takes up 11 bits and the tag length 5 bits, so we can merge them
+    // together.
+    let tag_len = (note_tag_len as u16) << 11;
+    let receiver_profile: u16 = tag_len | interface;
+    receiver_profile.to_be_bytes()
+}
+
+/// Reads the receiver profile from the provided bytes.
+fn decode_receiver_profile(
+    byte_iter: &mut impl ExactSizeIterator<Item = u8>,
+) -> Result<(AddressInterface, Option<u8>), AddressError> {
+    if byte_iter.len() < 2 {
+        return Err(AddressError::decode_error("expected two bytes to decode receiver profile"));
+    };
+
+    let byte0 = byte_iter.next().expect("byte0 should exist");
+    let byte1 = byte_iter.next().expect("byte1 should exist");
+    let receiver_profile = u16::from_be_bytes([byte0, byte1]);
+
+    let tag_len = (receiver_profile >> 11) as u8;
+    let note_tag_len = if tag_len == ABSENT_NOTE_TAG_LEN {
+        None
+    } else {
+        Some(tag_len)
+    };
+
+    let addr_interface = receiver_profile & 0b0000_0111_1111_1111;
+    let addr_interface = AddressInterface::try_from(addr_interface).map_err(|err| {
+        AddressError::decode_error_with_source("failed to decode address interface", err)
+    })?;
+
+    Ok((addr_interface, note_tag_len))
+}
+
+/// Append encryption key variant discriminant and key to the provided vector of bytes.
+fn encode_encryption_key(key: &SealingKey, encoded: &mut Vec<u8>) {
+    match key {
+        SealingKey::X25519XChaCha20Poly1305(pk) => {
+            encoded.push(ENCRYPTION_KEY_X25519_XCHACHA20POLY1305);
+            encoded.extend(&pk.to_bytes());
+        },
+        SealingKey::K256XChaCha20Poly1305(pk) => {
+            encoded.push(ENCRYPTION_KEY_K256_XCHACHA20POLY1305);
+            encoded.extend(&pk.to_bytes());
+        },
+        SealingKey::X25519AeadRpo(pk) => {
+            encoded.push(ENCRYPTION_KEY_X25519_AEAD_RPO);
+            encoded.extend(&pk.to_bytes());
+        },
+        SealingKey::K256AeadRpo(pk) => {
+            encoded.push(ENCRYPTION_KEY_K256_AEAD_RPO);
+            encoded.extend(&pk.to_bytes());
+        },
+    }
+}
+
+/// Reads the encryption key from the provided bytes.
+fn decode_encryption_key(
+    byte_iter: &mut impl ExactSizeIterator<Item = u8>,
+) -> Result<SealingKey, AddressError> {
+    // Need at least 1 byte for discriminant
+    if byte_iter.len() < 1 {
+        return Err(AddressError::decode_error(
+            "expected at least 1 byte for encryption key variant",
+        ));
+    };
+
+    // Read variant discriminant
+    let variant = byte_iter.next().expect("variant byte should exist");
```

Comment:
Nit: I'd write this as `let Some(key_discriminant) = byte_iter.next() else { return Err(...) }`, so we can remove the `expect` call.